### ISSUE TITLE
Transparency

### ISSUE THREAD

**Issue by LoganDark** [john01dav/softbuffer#17]
Right now, softbuffer seems to completely ignore the alpha channel. It would be nice if this was supported.

**Comment by i509VCB** [comment]
A transparency hint would be useful in the Wayland backend, as we could tell the compositor about the window content being opaque and let the compositor apply some optimizations for rendering.

**Comment by ids1024** [comment]
In addition to a way to set this, it would be good to have an API to test if it's supported. On Wayland RGBA and RGBX should always be, but I think this may not be true on others?

What kind of API would be suitable here? A `set_alpha()` method? `supports_alpha()`? Maybe there's a better way to name it.

This becomes a lot more complicated if we wanted to support formats other than RGBA/RGBX.

**Comment by cafce25** [comment]
It's probably by accident, but on windows 11 tranparency does work by writing the alpha value in the first byte which `Surface` specifies to always be 0:
```
let mut buffer = self.surface.buffer_mut()?;
let image_data = pixmap.data();
for index in 0..(width * height) as usize {
    let [r, g, b, a] = image_data[index * 4..][..4].try_into()?;
    buffer[index] = u32::from_be_bytes([a, r, g, b]);
}
```
It might be worth to preserve that behaviour if feasible.

**Comment by madsmtm** [comment]
On macOS/iOS, I think it's possible to query the transparency state of the surface, something like [`-[CALayer isOpaque]`](https://developer.apple.com/documentation/quartzcore/calayer/isopaque?language=objc)? If that's possible on other platforms too, maybe we could make this entirely transparent (haha) to the user?

If not, then I'd be in favour of a `fn set_transparent(&mut self, state: bool)` on `Surface`, which switches between `BGRX` and `BGRA` (or whatever pixel format we end up with).

**Comment by MarijnS95** [comment]
Android has [`ASurfaceTransaction_setBufferTransparency()`](https://developer.android.com/ndk/reference/group/native-activity#asurfacetransaction_setbuffertransparency) and [`ASurfaceTransaction_setBufferAlpha()`](https://developer.android.com/ndk/reference/group/native-activity#asurfacetransaction_setbufferalpha) which becomes relevant on my impending `ASurfaceControl`/`ASurfaceTransaction`-based Android backend for `softbuffer`.

For the "root" `ANativeWindow` (== Java `Surface`), I don't think that's allowed to be transparent (maybe some edge cases for floating windows with permissions?) but any surface implicitly layered on top via the view tree / Java (or aforementioned `ASurfaceControl`!) could be allowed to be transparent, if assigned an alpha format (`RGBA` as opposed to `RGBX`).

**Comment by madsmtm** [comment]
There's also the question of whether to use "straight" or "premultiplied" alpha? See [the W3C spec](https://html.spec.whatwg.org/multipage/canvas.html#premultiplied-alpha-and-the-2d-rendering-context) for a good description of the difference.

I think the Web Canvas API uses straight alpha, so if we want to be zero-copy there, we're forced to use straight alpha in softbuffer too?

Or maybe we provide a function like:
```rust
enum Transparency {
    Opaque,
    #[default]
    StraightAlpha,
    PremultipliedAlpha,
}

impl Surface {
    fn set_transparency(&mut self, value: Transparency) { todo!() }
}
```

And then only `Transparency::StraightAlpha` would be zero-copy on Web?

**Comment by madsmtm** [comment]
> Is there though? Premultiplied alpha is for game engines and compositors. Desktop windowing systems have always been straight alpha. I don't think there's a single mainstream platform that exposes premultiplied alpha for window surfaces. So why have it in softbuffer? Just for future proofing?

But I don't know the prior art here, only that [CoreGraphics allows you to to use premultiplied values](https://developer.apple.com/documentation/coregraphics/cgimagealphainfo/premultipliedfirst?language=objc), so the context is useful!

Or are you implicitly assuming that Softbuffer surfaces correspond 1 to 1 to windows? Because I think we should strive to not have that assumption, it should be possible to have a Softbuffer surface somewhere inside e.g. a deeply nested hierarchy of `NSView`s.

**Comment by madsmtm** [comment]
> Doesn't it use `HasRawWindowHandle`?

Yes, but that's really badly named, it should be more like `HasRawViewHandle` or `HasRawSomethingYouCanAddASurfaceToHandle` IIUC.
